### PR TITLE
Dispatch to use fp32 distance computation in NN Descent depending on data dimensions

### PR DESCRIPTION
**Pull request by jinsolp**
Closes https://github.com/rapidsai/cuvs/issues/1370
Closes https://github.com/rapidsai/cuvs/issues/195

This PR adds an option to use fp32 distance computation.

~~(Outdated) From heuristics, chose dim=16 as the threshold for dispatching to a fp32 distance kernel.
We do manual computation, but since we only target small dimensions, fp32 dispatching ends up being slightly faster end to end with much better recall for small dimensions.~~

All number below are run on L40 machine and AMD EPYC CPU with 128 cores. Perf and recall is averaged over 5 runs and all time is in seconds. Baseline knn graph is computed using `sklearn.neighbors.NearestNeighbors` brute for method.

### Max iters=20
Screenshot 2025-10-08 at 10 56 17 AM

For larger dimensions there is an inherent issue with the NN Descent algorithm itself that makes the recall low. This can be improved slightly with more iterations.
Also notice that the e2e time taken is similar or slightly less for using fp32.

### Max iters=100
Screenshot 2025-10-08 at 10 58 26 AM

Notice how the blue part, the recall doesn't get better compared to the table above even with more iterations (i.e. why we need the fp32 appraoch for this part)

# Perf impact on different architectures

### H100
Screenshot 2025-11-20 at 10 15 25 AM

### L40
Screenshot 2025-11-20 at 10 15 50 AM

### CODE REVIEW

**Comment by jinsolp** [comment]
Maybe we might need to have fp16 vs fp32 as an option (and default to fp16) instead of depending on data dimensions. WIP investigating

File: cpp/src/neighbors/detail/nn_descent.cuh [hunk 1]
```diff
@@ -476,6 +478,326 @@ __device__ __forceinline__ void remove_duplicates(
   }
 }
 
+template <typename Index_t, typename Data_t, typename DistEpilogue_t>
+__device__ __forceinline__ void calculate_metric(float* s_distances,
```
Comment (jinsolp) [inline cpp/src/neighbors/detail/nn_descent.cuh:486]: This is taken out as a separate function from the original kernel because it's commonly used for fp32 and fp16

File: cpp/src/neighbors/detail/nn_descent.cuh [hunk 2]
```diff
@@ -476,6 +478,326 @@ __device__ __forceinline__ void remove_duplicates(
...
+        int tmp_row = i / SKEWED_MAX_NUM_BI_SAMPLES;
+        int tmp_col = i % SKEWED_MAX_NUM_BI_SAMPLES;
+        if (tmp_row < list_new_size && tmp_col < list_new_size) {
+          float acc = 0.0f;
+          for (int d = 0; d < num_load_elems; d++) {
+            acc += s_nv[tmp_row][d] * s_nv[tmp_col][d];
+          }
+          s_distances[i] += acc;
+        }
+      }
```
Comment (jinsolp) [inline cpp/src/neighbors/detail/nn_descent.cuh:676]: This matmul part is different from the fp16 kernel

**Comment by jinsolp** [comment]
Thanks for the feedback @divyegala, changing this to target 26.02 for now because Corey suggested we do further investigation. Force-pushing after rebasing

File: cpp/src/neighbors/detail/nn_descent.cuh [hunk 3]
```diff
@@ -1039,6 +1316,20 @@ GNND<Data_t, Index_t>::GNND(raft::resources const& res, const BuildConfig& build
 {
   static_assert(NUM_SAMPLES <= 32);
 
+  using input_t = typename std::remove_const<Data_t>::type;
+  if (std::is_same_v<input_t, float> && build_config.dataset_dim <= 16) {
```
Comment (jinsolp) [inline cpp/src/neighbors/detail/nn_descent.cuh]: dispatch logic based on dimensions

**Comment by jinsolp** [comment]
Blue is better  (e2e time or kernel time)
- FP32 kernel is usually faster for small dimensions
- FP32 e2e time is sometimes slightly longer for small dimensions despite shorter kernel time. Believe this small gap comes from more data transfers (4bytes vs 2bytes per data element)

## L40 GPU
Screenshot 2025-12-02 at 6 28 48 PM
Screenshot 2025-12-02 at 6 28 58 PM

## H100 GPU
Screenshot 2025-12-02 at 6 30 47 PM
Screenshot 2025-12-02 at 6 30 59 PM

**Comment by jinsolp** [comment]
@divyegala based on the latest benchmarks (and confirming from HDBSCAN that we don't have quality issues from using fp16 for larger dimensions), the mechanism is back to dispatching based on the number of dimensions 
(chose 16 as a threshold).

> I would like to see a unit test, if possible, where you can demonstrably prove distance computation improvements in fp32 over fp16.

Not sure how to add this as a test. Any suggestions?

**Comment by divyegala** [comment]
> Not sure how to add this as a test. Any suggestions?

@jinsolp is it possible to generate a dataset that shows degraded recall for fp16 and good recall for fp32? The way I envision the assertion is `ASSERT(fp16_recall < <upperbound_value> && fp32_recall >= <lowerbound_value>)`.

**Comment by jinsolp** [comment]
We can (e.g. the blobs data with dim=2 which shows 0.4 recall for fp16 vs 0.99 recall for fp32)
, but if we run through the current code, it will always take the fp32 path.

Should I hardwire the results of fp16 and add that as a vector to the test?

**Comment by divyegala** [comment]
@jinsolp we should allow the user an override in-case the heuristic fails. Let users choose fp32 explicitly if they want, and exporting that option will also let you test the separate paths.

**Comment by jinsolp** [comment]
Oh okay, so let the user choose an option, and if none is given fall back to the heuristic dim=16 threshold?

**Comment by divyegala** [comment]
The other way around. Use the heuristic eagerly, but let user override.

**Comment by jinsolp** [comment]
@divyegala we now have a `dist_comp_dtype` option defaulting to `auto`. can override by using `fp32` or `fp16`.
Added a test as well.

File: c/src/neighbors/nn_descent.cpp
Comment (divyegala) [inline c/src/neighbors/nn_descent.cpp:1]: Please default initialize the index creation with this new parameter.

File: cpp/include/cuvs/neighbors/nn_descent.hpp [hunk 4]
```diff
@@ -37,13 +47,19 @@ namespace cuvs::neighbors::nn_descent {
...
+ * valid when data type is fp32. Use `FP16` for better performance and memory usage at the cost of
+ * precision.
  */
 struct index_params : cuvs::neighbors::index_params {
   size_t graph_degree              = 64;
   size_t intermediate_graph_degree = 128;
   size_t max_iterations            = 20;
   float termination_threshold      = 0.0001;
   bool return_distances            = true;
+  DIST_COMP_DTYPE dist_comp_dtype  = DIST_COMP_DTYPE::AUTO;
```
Comment (jinsolp) [inline cpp/include/cuvs/neighbors/nn_descent.hpp:62]: @divyegala It already defaults to AUTO here, which would be taken in `nn_descent.cpp`

Comment (divyegala) [reply cpp/include/cuvs/neighbors/nn_descent.hpp:62]: That file is the source of C API, and this default is for the C++ API.

Comment (jinsolp) [reply cpp/include/cuvs/neighbors/nn_descent.hpp:62]: I believe all the defaults of cpp are passed when we call `cuvsNNDescentIndexParamsCreate`
https://github.com/rapidsai/cuvs/blob/1959a0dda1f1d1e49124860f1a766793e1598706/c/src/neighbors/nn_descent.cpp#L167-L182

C doesn't support default struct init and so nothing is initialized in the c struct:
https://github.com/rapidsai/cuvs/blob/1959a0dda1f1d1e49124860f1a766793e1598706/c/include/cuvs/neighbors/nn_descent.h#L38-L46

Comment (divyegala) [reply cpp/include/cuvs/neighbors/nn_descent.hpp:62]: I haven't kept up with the changes to the C API, my apologies. We used to explicitly default initialize in the create function previously.
